wgsl/index.bs

```diff
@@ -7347,13 +7345,344 @@ See [[#expressions]].
+This way, whenever a function is analyzed, the metadata for all of its callees has already been computed.
+There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+Note: another way of saying the same thing is that we do a topological sort of functions ordered by the "is a (possibly indirect) callee of" partial order, and analyze them in that order.
+
+### Analyzing the uniformity requirements of a function ### {#uniformity-function}
+
+Each function is analyzed in two phases.
+
+The first phase walks over the syntax of the function, building a directed graph along the way based on the rules in the following subsections.
+The second phase explores that graph, resulting in either rejecting the program, or computing the constraints on calling this function.
+
+<div class="note">Note: apart from two special nodes MustBeUniform and CannotBeUniform, all nodes can be understood as having one of the following meanings:
+        - The control flow at a specific program point must be uniform
+        - The value of a specific expression must be uniform
+        - A specific variable must always hold a uniform value
+
+        An edge can be understood as an implication from the statement corresponding to its source node to the statement corresponding to its target node.
+
+        To express that something must always be uniform (e.g. the control flow at the call site of a derivative), we add an edge from MustBeUniform to the corresponding node.
+        One way to understand this, is that MustBeUniform corresponds to the proposition True, so that MustBeUniform -> X is the same as saying that X is true.
+
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we add an edge from the corresponding node to CannotBeUniform.
+        One way to understand this, is that CannotBeUniform corresponds to the proposition False, so that X -> CannotBeUniform is the same as saying that X is false.
+
+        A consequence of this interpretation is that every node reachable from MustBeUniform corresponds to something which must be uniform for the program to be valid, and every node from which cannotBeUniform is reachable corresponds to something whose uniformity we cannot guarantee. It follows that we have a uniformity violation (and thus reject the program) if there is any path from MustBeUniform to CannotBeUniform.
+</div>
+
+In more detail, here is what we compute for each function:
+
+* A tag about the control-flow in which the function can be called: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter}
+* A tag for each parameter of the function: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter, MustBeUniformForReturnValue, NoRestriction}
```

Comment:
It sounds like exactly one tag needs to be attached to a node representing the function parameter. Would it ever be the case where we want `MustBeUniformForControlFlowAfter` together with `MustBeUniformForReturnValue` but not as strong as `AlwaysMustBeUniform`?